PluginComponent contributes description to -Xshow-phases. (Fixes SI-6446)

som-snytt opened the pull request:
In Global, SubComponent is called a phase descriptor, but it doesn't actually have a description. (Phase itself does.) This fix adds a description to PluginComponent so that plugins can describe what they do in -Xshow-phases.

To avoid chaos, as per the formatting comments in Global.scala, -Xshow-phases output is restricted to 80 columns.  As an aid, inlineExceptionHandlers is renamed to inlineHandlers.

Phase descriptions are provided for the CPS components.

A minimal test is provided for -Xshow-phases with a custom plugin.  Plugin loading and partest are modified so that a test can provide a plugin.  (Plugin loading accepts an exploded jar, and partest replaces "dot" or cwd with output dir.  This means a test can contribute just one plugin.)

(One could imagine testing a compiler component by supplying a test phase that creates possibly screwy trees, for example to test robustness.)

Moves test run/t4351.scala from the pos test set.

Footnote: Continuations tests could be mainstreamed, by adding .flags.

review by: @paulp who cares

som-snytt committed:
PluginComponent contributes description to -Xshow-phases.

In Global, SubComponent is called a phase descriptor, but it doesn't
actually have a description.  (Phase itself does.)  This fix adds
a description to PluginComponent so that plugins can describe what
they do in -Xshow-phases.

Elliptical descriptions

Exploded archives

Plugged-in partest

Roundup at the Little h!

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/748/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1458/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1458/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/748/

paulp commented on test/files/run/programmatic-main.check:
```diff
@@ -1,31 +1,31 @@
+       cleanup  21  platform-specific cleanups, generate reflective calls
+         icode  22  generate portable intermediate code
+       inliner  23  optimization: do inlining
+inlineHandlers  24  optimization: inline exception handlers
```
That "H" is now the only big boy letter among all the phase names.

som-snytt replied:
Wow, you really do care!  Good catch.  Luckily, "h" is also tall, so it should be legible despite length.

som-snytt replied:
I was sanguine at first, but I'd forgotten that the string is scattered about.  My original edit was to override phaseName in global; but maybe with lazy val phase objects, there'd be a problem at assembly; second idea was just an enum of phase names and descriptions, so at least they're all in one place.  But I haven't read all the comments in firstPhase yet.

I didn't even realize inlineHandlers is a setting.  I can't be trusted.

paulp commented on src/compiler/scala/tools/nsc/plugins/PluginDescription.scala:
```diff
@@ -45,31 +40,62 @@ abstract class PluginDescription {
+  }
+
+/*
+  import scala.reflect.runtime.{currentMirror => cm, universe => uni}
```
In the interests of keeping the number of aliases to a minimum, I believe the standard for universe is "ru".

som-snytt replied:
Right. That is in a comment that answers, is this easier to do reflectively, and the answer is not really.  (You might remember someone asked on the ML can I avoid using string names when constructing objects from amorphous data.)  Shall I just delete the comment?

The fun part of that trial was supplying ctor defaults like PluginMain for the class, but that was just a lark.  No one is clamoring to reduce boilerplate for plugins.

som-snytt commented:
There are additional cogent rejoinders buried behind the outdated diffs.  I forgot that part about pushing.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/767/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1478/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/767/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1478/

som-snytt commented:
This is the partestable version of 
https://github.com/scala/scala/issues/1419

If it's not deemed to pull its weight, I'll kill it before it's no longer mergeable.  I already see issues with the issue.

@gkossakowski or @paulp might ask for a more substantial contribution to consume review-by resources.

rtyley committed:
Add convenience attribute operator to NodeSeq

Compared to the current method of reading the string text of an
attribute:

(x \ "@bar").text

...the new operator removes the need for a pair of parenthesis and
shortens the overall expression by 7 chars :

x \@ "bar"



Discussion on scala-internals:

https://groups.google.com/d/topic/scala-internals/BZ-tfbebDqE/discussion

adriaanm committed:
Merge pull request #1679 from rtyley/nodeseq-attribute-operator

Add convenience attribute operator to NodeSeq

adriaanm committed:
Merge pull request #1765 from paulp/pr/remove-detach

Removing more of that which merits removal

som-snytt commented:
Reopening with rebased commit.  Writing a plugin is so much more fun with a custom description.  I understand that it may be a slow process, but worth the wait.  Also, quite a lot of elven labor frees up in about a week, so maybe it will get some cheer.  (I wasn't actually referring to the elves who appear in the workshop every night to work on 2.10-final, but maybe they could help, too.)

I added more tests and tweaked some code. It's too bad the formatting logic is in global and not in reporter.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1910/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1910/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1194/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1194/

paulp commented:
This looks great to me.

paulp commented:
"It's too bad the formatting logic is in global and not in reporter."

We have the power to move stuff.

som-snytt commented:
I remember looking at moving it, and I've forgotten the detail, but it had to do with how the "info" return from compiler interacts.  (Or I don't remember.) So if desired, I would refactor that separately.

som-snytt commented:
Thanks.  I noticed the sbt stack trace with the plugin classloading issue didn't include this commit (whew?).  My New Year's resolution is to always have a PR in the review queue.  Excuse the split infinitive.  Happy Holidays.
